src/background.js

```diff
@@ -1,10 +1,15 @@
 window.scrNoti = window.scrNoti || {};
+const seenMessages = {};
+
+// TODO:
+// onQuit
```

Comment:
How would I intercept the closing down of Thunderbird, so that we can send an "quit" (or "end") message to the script?

Reply:
I don't think it is possible. And anyway Thunderbird can crash, etc. 

Why would you want such event by the way? I'm not sure it would be really useful.

Reply:
The script should know, that Thunderbird ends. For example it could remove any remaining notifications or stop a long-running process etc.

Yes, if Thunderbird crashes, the script would not know it. But then it is a crash and you would expect some side-effects in that case.

Reply:
A way I can think of would be some kind of continous loop checking for something in Thunderbird indicating it is currently being closed. I don't know if such thing exists and it would probably not be reliable.

The only viable solution I see, if you _really_ need such "Thunderbird has exited" hook in your script, would be for the add-on to sent a ping to the script every 5 seconds or so. If the script hasn't received a ping for, let's say 30 seconds, it can assume Thunderbird is not running.

But I don't know, that feels very hackish to me. Maybe that could be an option and not hardcoded.

Reply:
Amazing, that the `onQuit` callback is not exposed!

In the meantime I would suggest to include the PID of the thunderbird process into the payload. A script could then check for this PID if thus inclined.

Do you know, how to get the PID of the Thunderbird's main process?

Reply:
I guess you would have to call an external script in order to get that PID! ;-)

But of course the script itself could check for the existence of a running "Thunderbird" process , no need for the PID from the add-on after all..

Reply:
I just remembered that the add-on [Send Later](https://addons.thunderbird.net/en-us/thunderbird/addon/send-later-3/) shows a notification, when you want to quit Thunderbird if a message has not been sent.

These are the relevant lines: https://github.com/Extended-Thunder/send-later/blob/7f346e9c08de5a2f6b4854fc2978d8eec34fc5ac/background.js#L93-L109

It uses `messenger.quitter`, but I could not find it in the [official API documentation](https://webextension-api.thunderbird.net/en/stable/). Could this be a solution?

Reply:
It seems that they have access to this because they are using an "experiment api" :
https://github.com/Extended-Thunder/send-later/blob/master/manifest.json#L123

If you look at the commit history for "Scriptable Notifications", I was asked by the reviewer from Thunderbird team it is was possible for me to remove any usage of "experiment apis", which I did by implementing "native messaging" instead!

I don't want this extension to use such API again... Web Extensions is where things are going. Their APIs will improve over time. 

Sadly, we will have to live without any "quit" event if we can't find a way to implement it without an experiment api.

Reply:
Also, I think this experiment addon is used _to quit_ Firefox/Thunderbird, not to get an event when is is closing:

https://github.com/mozilla-extensions/quitter

Reply:
Ah, I see. Thanks.

No, we don't want to pollute the API with experimental features. If it ever becomes part of the standard api, we could include it. Should we open an issue with that regard?

Reply:
If you want! I think that feature is not that important.